history-traversal.md

```diff
@@ -0,0 +1,107 @@
+# History traversal
+
+These rough algorithms describe how portal activation, adoption, reportaling, and deportaling should work.
+
+Eventually, these steps will be moved into the portal spec, and the history traversal section of the HTML spec.
+
+## Definitions
+
+<dl>
+  <dt>Navigable</dt>
+  <dd>something that has session history, such as a top level page, an iframe, or a portal (although a portal can only have one item of session history).</dd>
+</dl>
+
+## Activate a portal in a 'push' style
+
+This is a regular activation that clears any 'forward' items in join session history and adds a new top-level history entry for the portaled document.
+
+1. Let _document_ be the portal's session history item's document.
+1. Let _targetHistoryItem_ be a copy of the portal's session history item.
+1. Remove _targetHistoryItem_'s document.
+1. Give _targetHistoryItem_ a weak reference to _document_.
```

Comment:
Is this actually easier/clearer than having the history item has a strong/weak flag rather than two separate fields?

Reply:
This isn't as easy as I thought.

Good bits:

- Avoids moving the document from property to property
- Avoids cases where the weak document and main document are set at the same time (and are different)

Less good bits:

- Maybe the moving solidifies the 'documents move around model' that we agreed on?
- It doesn't work with portals. With portals the history item is removed upon activation, so the weak reference is on the portal itself. We could keep the history item and set the weak flag there, but that means any portal rendering/interaction code would need to treat the history entry as absent if the weak flag is set. Although, this would ensure the weak reference is gone when the portal is navigated.

There may be other places where we'd need to avoid doing things with the session history item if the document reference is weak.

I'm undecided.